memiavl/tree.go

```diff
@@ -16,47 +18,82 @@ type Tree struct {
 	version uint32
 	// root node of empty tree is represented as `nil`
 	root Node
-
+	// write ahead log to store changesets for each block
+	bwal           *blockWAL
 	initialVersion uint32
 }
 
 // NewEmptyTree creates an empty tree at an arbitrary version.
-func NewEmptyTree(version int64) *Tree {
+func NewEmptyTree(version uint64, pathToWAL string) (*Tree, error) {
 	if version >= math.MaxUint32 {
 		panic("version overflows uint32")
 	}
-	return &Tree{version: uint32(version)}
+
+	var (
+		wal *blockWAL
+		err error
+	)
+	if pathToWAL != "" {
+		wal, err = newBlockWAL(pathToWAL, version, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &Tree{version: uint32(version), bwal: wal}, nil
 }
 
 // New creates an empty tree at genesis version
-func New() *Tree {
-	return NewEmptyTree(0)
+func New(pathToWAL string) (*Tree, error) {
+	return NewEmptyTree(0, pathToWAL)
 }
 
 // New creates a empty tree with initial-version,
 // it happens when a new store created at the middle of the chain.
-func NewWithInitialVersion(initialVersion int64) *Tree {
+func NewWithInitialVersion(initialVersion int64, pathToWAL string) (*Tree, error) {
 	if initialVersion >= math.MaxUint32 {
-		panic("version overflows uint32")
+		return nil, errors.New("version overflows uint32")
+	}
+	tree, err := New(pathToWAL)
+	if err != nil {
+		return nil, err
 	}
-	tree := New()
 	tree.initialVersion = uint32(initialVersion)
-	return tree
+	return tree, nil
 }
 
 // NewFromSnapshot mmap the blob files and create the root node.
-func NewFromSnapshot(snapshot *Snapshot) *Tree {
+func NewFromSnapshot(snapshot *Snapshot, pathToWAL string) (*Tree, error) {
 	if snapshot.IsEmpty() {
-		return NewEmptyTree(int64(snapshot.Version()))
+		return NewEmptyTree(uint64(snapshot.Version()), pathToWAL)
 	}
+
+	var (
+		wal *blockWAL
+		err error
+	)
+	if pathToWAL != "" {
+		wal, err = newBlockWAL(pathToWAL, uint64(snapshot.Version()), nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	return &Tree{
 		version: snapshot.Version(),
 		root:    snapshot.RootNode(),
-	}
+		bwal:    wal,
+	}, nil
 }
 
 // ApplyChangeSet apply the change set of a whole version, and update hashes.
-func (t *Tree) ApplyChangeSet(changeSet *iavl.ChangeSet, updateHash bool) ([]byte, int64, error) {
+// Returns hash, new version, and potential error.
+func (t *Tree) ApplyChangeSet(changeSet iavl.ChangeSet, updateHash bool) ([]byte, int64, error) {
```

Comment:
yes, pointer was wrong [use](https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9), that var will be reused in for loop